[TA] Client side validation

mssfang opened the pull request:
Added the v-team shared client-side validation

kristapratico commented on sdk/textanalytics/azure-ai-textanalytics/src/main/java/com/azure/ai/textanalytics/AnalyzeActionsAsyncClient.java:
```diff
@@ -174,33 +175,43 @@ class AnalyzeActionsAsyncClient {
             final boolean finalIncludeStatistics = options.isIncludeStatistics();
 
             if (service != null) {
+                throwIfLegacyApiVersionForActions(actions,
```
I might be reading the code wrong, but if service != null then doesn't that mean this is the language API client? Do we still need to the do the validation for 3.0 and 3.1 here?

mssfang replied:
You are right. This is a redundant code that it never called

kristapratico reviewed (commented):
In Python and .NET we also updated our docstrings to indicate version support. Is this something Java does?

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[azure-ai-textanalytics](https://apiview.dev/Assemblies/Review/8dc3d5c0e41e46578be7e4fbcadc9dc5)

mssfang commented:
> In Python and .NET we also updated our docstrings to indicate version support. Is this something Java does?

Yeah. Java also should update the docstring. Added the docstring.

heaths commented on sdk/textanalytics/azure-ai-textanalytics/src/main/java/com/azure/ai/textanalytics/TextAnalyticsAsyncClient.java:
```diff
@@ -300,6 +306,9 @@ public Mono<DetectedLanguage> detectLanguage(String document, String countryHint
      * @return A {@link Mono} contains a {@link DetectLanguageResultCollection}.
      *
      * @throws NullPointerException if {@code documents} is null.
+     * @throws IllegalStateException if {@link TextAnalyticsRequestOptions#isServiceLogsDisabled()} is true in service
```
I forgot to do this for .NET. Thanks for the reminder.

heaths reviewed (approved):
Did you also want to add the `serviceLogsDisabled` "property" to the newer actions like I did in Azure/azure-sdk-for-net#30753 as well? Makes everything consistent; though, could be another PR. No need to validate those, though, since they only exist for 2022-05-01 and newer anyway.

kristapratico commented on sdk/textanalytics/azure-ai-textanalytics/src/main/java/com/azure/ai/textanalytics/AnalyzeActionsAsyncClient.java:
```diff
@@ -174,26 +175,33 @@ class AnalyzeActionsAsyncClient {
         try {
             Objects.requireNonNull(actions, "'actions' cannot be null.");
+            throwIfTargetServiceVersionFound(this.serviceVersion, Arrays.asList(TextAnalyticsServiceVersion.V3_0),
+                "'beginAnalyzeActions' is only available for API version v3.1 and up.");
```
Agreed, will open a PR updating to that messaging.

mssfang commented on sdk/textanalytics/azure-ai-textanalytics/src/main/java/com/azure/ai/textanalytics/AnalyzeActionsAsyncClient.java:
```diff
@@ -229,6 +237,8 @@ PollerFlux<AnalyzeActionsOperationDetail, AnalyzeActionsResultPagedFlux> beginAn
             }
 
+            throwIfTargetServiceVersionFoundForActions(this.serviceVersion,
+                Arrays.asList(TextAnalyticsServiceVersion.V3_0, TextAnalyticsServiceVersion.V3_1), actions);
```
The logic of `throwIfTargetServiceVersionFoundForActions` is if found the API source version which the client is using in the target version list, we will throw an exception. So the target version is the version that the api/property should not call.
It won't be long I believe, since we only GA 2 GA versions in the past 3 years, and the older version might get deprecated at some time in the future when the service decided to not support it. It should be fine.

mssfang commented on sdk/textanalytics/azure-ai-textanalytics/src/main/java/com/azure/ai/textanalytics/AnalyzeActionsAsyncClient.java:
```diff
@@ -1284,4 +1298,27 @@ private String[] parseActionErrorTarget(String targetReference, String errorMess
+
+    private void throwIfTargetServiceVersionFoundForActions(TextAnalyticsServiceVersion sourceVersion,
+        List<TextAnalyticsServiceVersion> targetVersions, TextAnalyticsActions actions) {
+        if (actions.getMultiLabelClassifyActions() != null) {
```
The default value for those actions should be `null`. If the collection is not `null`, it implies the customer is trying to use it and then we need to do the client-side validation to check if the service API is valid.

mssfang commented on sdk/textanalytics/azure-ai-textanalytics/src/main/java/com/azure/ai/textanalytics/implementation/Utility.java:
```diff
@@ -1348,4 +1349,15 @@ public static SentenceAssessment findSentimentAssessment(String assessmentPointe
     }
+
+    public static void throwIfTargetServiceVersionFound(TextAnalyticsServiceVersion sourceVersion,
+                                                        List<TextAnalyticsServiceVersion> targetVersions,
```
It is a static variable, so it will be one object(eg, V3.0) for TextAnalyticsServiceVersion class. It won't create a new object TextAnalyticsServiceVersion.V3_0 for each call.

deyaaeldeen commented on sdk/textanalytics/azure-ai-textanalytics/src/main/java/com/azure/ai/textanalytics/implementation/Utility.java:
```diff
@@ -1348,4 +1349,37 @@ public static SentenceAssessment findSentimentAssessment(String assessmentPointe
+        for (TextAnalyticsServiceVersion targetVersion : targetVersions) {
+            if (targetVersion != null && sourceVersion != null
+                && targetVersion.getVersion().equals(sourceVersion.getVersion())) {
+                throw LOGGER.logExceptionAsError(new IllegalStateException(errorMessage));
```
Am I reading this correctly that it throws in every loop iteration? Shouldn't this function throw just once with an error message that has all the target service versions?

deyaaeldeen replied:
Never mind, I just looked at how it is being used again and it will throw at max only once since the input service version can only match one of the input target versions.

mssfang replied:
In java. the uncaught exception will terminate the thread. so this for loop will not continue to run. The method will exit immediately.

deyaaeldeen commented on sdk/textanalytics/azure-ai-textanalytics/src/main/java/com/azure/ai/textanalytics/implementation/Utility.java:
```diff
@@ -1348,4 +1349,37 @@ public static SentenceAssessment findSentimentAssessment(String assessmentPointe
+     */
+    public static String getUnsupportedServiceApiVersionMessage(String unsupportedName,
+        TextAnalyticsServiceVersion sourceVersion, TextAnalyticsServiceVersion minSupportedVersion) {
+        return String.format("'%s' is not available in API version %s. Use service API version '%s' or newer.",
```
Is it clear to the customer that v2022-05-01 is newer than v3.1? Should the message be clearer in that regard?

mssfang replied:
I think we are suggesting using the min version of 2022-05-01. It is inclusive. Customers will find out which version is newer from CHANGELOG history, README, and the order of TextAnalyticsServiceVersion.

deyaaeldeen replied:
> I think we are suggesting using the min version of 2022-05-01

If the customer is trying to use `disableServiceLogs` with v3.0, would the error message suggest using v3.1 and newer or v2022-05-01 and newer?

mssfang replied:
for `disableServiceLogs`,  we would suggest the customer to use v3.1 at least. 
https://github.com/Azure/azure-sdk-for-java/pull/30708/files#diff-e6b45745e42ad20893ffc77376d77776b9e455cfa18486a3d6eee01676072804R186

deyaaeldeen commented on sdk/textanalytics/azure-ai-textanalytics/src/main/java/com/azure/ai/textanalytics/AnalyzeSentimentAsyncClient.java:
```diff
@@ -146,4 +155,20 @@ private Mono<Response<AnalyzeSentimentResultCollection>> getAnalyzedSentimentRes
+                getUnsupportedServiceApiVersionMessage("AnalyzeSentimentOptions.includeOpinionMining",
+                    serviceVersion, TextAnalyticsServiceVersion.V3_1));
+        }
+        if (options.isServiceLogsDisabled()) {
+            throwIfTargetServiceVersionFound(this.serviceVersion, Arrays.asList(TextAnalyticsServiceVersion.V3_0),
+                getUnsupportedServiceApiVersionMessage("TextAnalyticsRequestOptions.disableServiceLogs",
+                    serviceVersion, TextAnalyticsServiceVersion.V3_1));
+        }
```
should `throwIfCallingNotAvailableFeatureInOptions` be used here?

mssfang replied:
Actually this method should be renamed to the private method `throwIfCallingNotAvailableFeatureInOptions` in `AnalyzeSentimentAysncClient`. The reason to keep them separate independent methods in each xxxAsycnClient class is to support future new options properties.
